Marlin/src/HAL/STM32/timers.cpp

```diff
@@ -31,9 +31,10 @@
 // The TONE timer is not present here, as it currently cannot be set programmatically. It is set
 // by defining TIM_IRQ_PRIO in the variant.h or platformio.ini file, which adjusts the default
 // priority for STM32 HardwareTimer objects.
-#define SWSERIAL_TIMER_IRQ_PRIO_DEFAULT  1 // Requires tight bit timing to communicate reliably with TMC drivers
-#define SERVO_TIMER_IRQ_PRIO_DEFAULT     1 // Requires tight PWM timing to control a BLTouch reliably
-#define STEP_TIMER_IRQ_PRIO_DEFAULT      2
+// Let the serial interrupt have the highest priority (1) to avoid lost characters.
+#define SWSERIAL_TIMER_IRQ_PRIO_DEFAULT  2 // Requires tight bit timing to communicate reliably with TMC drivers
+#define SERVO_TIMER_IRQ_PRIO_DEFAULT     2 // Requires tight PWM timing to control a BLTouch reliably
+#define STEP_TIMER_IRQ_PRIO_DEFAULT      8
```

Comment:
This changes will be applied for every single board using STM32.

The old values looks better too me.. 🤔

Reply:
SWSERIAL is used by uart steppers. So I think is good for us it having a higher priority.

Step timer should not get postponed by another timer too, to not lose steps.

And servo, we have a long history of servo timing issues...

And, for normal uart, host can use CRC and resend. And even in high priority, some loss is expected when using uart. So I don't think we need prioritize uart over other timers.

And some board uses usb serial, that is lossless...

Do you have any reason to modify this timers priority?

The old values don't work for you?

Reply:
I was having a hard time with serial on STM32. I was using 921600 on STM32F1 without any issue. On STM32 anything >= 57600 causes a few characters to be lost randomly, the faster the baudrate is the worse it gets. stm32duino implements serial via interrupts and according to a few articles it indeed may lose data on 115200 or higher baudrates.

I'm now running with 57600 but it would still cause some corruption now and then. Resends always resolved the issue but I think it's better to have a stable connection. When I tweaked the interrupt priorities the serial connection is now solid and I couldn't find any other side effects, print quality is the same.

This is not ideal and I agree this is more of a hack than a fix. It seems the better approach is to read the serial with DMA (https://github.com/MaJerle/stm32-usart-uart-dma-rx-tx). I tried to do that but I'm not familiar with STM32 framework and eventually ended up with this workaround.

By the way, I'm well aware that Marlin cannot transmit data anywhere near this fast and 921600 is overkill.

Reply:
I was talking with @sjasonsmith about the timer priorities. He said that these values seems ok. He warned that maybe we should keep swserial and servo at 1, because of bltouch and uart steppers. 
Another issue that he pointed out is that we may get some issue running step timer in lower priority than usb serial (default is 1 and 5 when using usb flash drive).

This are the STM32F1 defaults:
```cpp
#define STEP_TIMER_IRQ_PRIO 2
#define TEMP_TIMER_IRQ_PRIO 3
#define SERVO0_TIMER_IRQ_PRIO 1
```
From this PR (that seems to solve a servo timer issue): https://github.com/MarlinFirmware/Marlin/pull/18637

Maybe we should discuss this timer priority in another PR/issue, because they impact in too many things...

Reply:
STM32F1 (aka libmaple core) is leaner than STM32 ('official' core), libmaple is nearly bare metal and hence lower overheads. however, libmaple is nearly strictly stm32f103. STM32 strives for compatibility across the lines and hence needs to wrap around STM HAL as that's about the only way to 'cross mcu'. tweaking irq priorities may have some implications that we'd not know for now, it may cause issues with stepping or servos.

try to go for boards that use usb-serial (usb-cdc-acm). stm32f103 has usb-serial as well it is mainly a hardware board design thing and it does not really make sense to use a separate usb-uart bridge to talk to stm32f103 when stm32f103 has a better native usb hardware device support.